utils/s2n_random.c

```diff
@@ -128,65 +129,73 @@ S2N_RESULT s2n_get_mix_entropy(struct s2n_blob *blob)
     return S2N_RESULT_OK;
 }
 
-void s2n_on_fork(void)
-{
-    zero_if_forked = 0;
-}
-
-static inline S2N_RESULT s2n_defend_if_forked(void)
+static S2N_RESULT s2n_initialise_drbgs(void)
 {
     uint8_t s2n_public_drbg[] = "s2n public drbg";
     uint8_t s2n_private_drbg[] = "s2n private drbg";
-    struct s2n_blob public = {.data = s2n_public_drbg,.size = sizeof(s2n_public_drbg) };
-    struct s2n_blob private = {.data = s2n_private_drbg,.size = sizeof(s2n_private_drbg) };
+    struct s2n_blob public = {.data = s2n_public_drbg, .size = sizeof(s2n_public_drbg) };
+    struct s2n_blob private = {.data = s2n_private_drbg, .size = sizeof(s2n_private_drbg) };
 
-    if (zero_if_forked == 0) {
-        /* Clean up the old drbg first */
-        RESULT_GUARD(s2n_rand_cleanup_thread());
-        /* Instantiate the new ones */
-        RESULT_GUARD(s2n_drbg_instantiate(&per_thread_public_drbg, &public, S2N_AES_128_CTR_NO_DF_PR));
-        RESULT_GUARD(s2n_drbg_instantiate(&per_thread_private_drbg, &private, S2N_AES_128_CTR_NO_DF_PR));
-        zero_if_forked_ptr = zeroed_when_forked_page;
-        zero_if_forked = 1;
-    }
+    RESULT_GUARD(s2n_drbg_instantiate(&s2n_per_thread_rand_state.public_drbg, &public, S2N_AES_128_CTR_NO_DF_PR));
+    RESULT_GUARD(s2n_drbg_instantiate(&s2n_per_thread_rand_state.private_drbg, &private, S2N_AES_128_CTR_NO_DF_PR));
+
+    s2n_per_thread_rand_state.drbgs_initialised = true;
 
     return S2N_RESULT_OK;
 }
 
-S2N_RESULT s2n_get_public_random_data(struct s2n_blob *blob)
+static S2N_RESULT s2n_maybe_initialise_drbgs(void)
 {
-    RESULT_GUARD(s2n_defend_if_forked());
+    if (s2n_per_thread_rand_state.drbgs_initialised == false) {
+        RESULT_GUARD(s2n_initialise_drbgs());
 
-    uint32_t offset = 0;
-    uint32_t remaining = blob->size;
+        /* Then cache the fork generation number. We just initialised the drbg
+         * states with new entropy and forking is not an external event.
+         */
+        uint64_t returned_fgn = 0;
+        RESULT_GUARD(s2n_get_fork_generation_number(&returned_fgn));
+        s2n_per_thread_rand_state.cached_fgn = returned_fgn;
+    }
 
-    while(remaining) {
-        struct s2n_blob slice = { 0 };
+    return S2N_RESULT_OK;
+}
 
-        RESULT_GUARD_POSIX(s2n_blob_slice(blob, &slice, offset, MIN(remaining, S2N_DRBG_GENERATE_LIMIT)));;
+/* s2n_defend_against_ube() implements defenses against ube's (uniqueness
+ * breaking events). Currently, only implements fork detection.
+ */
+static S2N_RESULT s2n_defend_against_ube(void)
+{
+    uint64_t returned_fgn = 0;
+    RESULT_GUARD(s2n_get_fork_generation_number(&returned_fgn));
 
-        RESULT_GUARD(s2n_drbg_generate(&per_thread_public_drbg, &slice));
+    if (returned_fgn != s2n_per_thread_rand_state.cached_fgn) {
```

Comment:
Should we add some kind of atomic_equals and atomic_notequals macro? It really matters that this check is atomic, right? On some platforms we might need compiler hinters or even asm to make it so.

Reply:
If these values were shared it would matter. But they are all thread-local, so we are safe from any threading issues. The value managed by `s2n_get_fork_generation_number()` is shared. But the function contract (and what is implemented) ensures that the value is incremented after a fork event.

Finally, an (process-)external event shouldn't be able to fork the process running this code. If this was possible, we would be in trouble here. In Snapsafe this is an issue though...